crates/jstz_core/src/kv/outbox.rs

```diff
@@ -0,0 +1,461 @@
+    }
+
+    pub fn queue_message(&mut self, message: OutboxMessage) {
+        self.0.push(message)
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OutboxQueueMeta {
+    /// Current queue length between the rollup and snapshots'
+    /// outbox queues
+    queue_len: u32,
+    /// Maximum capacity of the rollup outbox queue
+    max: u32,
+}
+
+impl OutboxQueueMeta {
+    // FIX: Unfortunately, the RollupOutboxQueue does not expose any methods to
+    // read its metadata so use this workaround for now.
+    pub fn read_meta(rt: &impl Runtime) -> Result<Option<OutboxQueueMeta>> {
+        Storage::get::<OutboxQueueMeta>(rt, &OUTBOX_QUEUE_ROOT)
+    }
+
+    fn write_meta(&self, rt: &mut impl Runtime) -> Result<()> {
+        Storage::insert(rt, &OUTBOX_QUEUE_ROOT, self)
+    }
+}
+
+/// An outbox queue that composes outbox queue snapshots
+/// and `ROLLUP_OUTBOX_QUEUE` while maintaining the invariants
+/// of the kernel outbox.
+#[derive(Debug)]
+pub struct OutboxQueue {
+    /// Metadata for the outbox queue
+    meta: OutboxQueueMeta,
+
+    /// Rollup outbox queue
+    rollup_outbox_queue: RollupOutboxQueue,
+}
+
+impl OutboxQueue {
+    fn write_outbox_message(
+        rt: &mut impl Runtime,
+        message: &OutboxMessageFull<impl AtomicBatch>,
+    ) -> Result<()> {
+        let mut buffer = Vec::with_capacity(MAX_OUTPUT_SIZE);
+        message
+            .bin_write(&mut buffer)
+            .map_err(|_| OutboxError::OutboxMessageSerializationError)?;
+        rt.write_output(&buffer)?;
+        Ok(())
+    }
+
+    pub fn new(max: u32) -> Self {
+        Self {
+            meta: OutboxQueueMeta { queue_len: 0, max },
+            rollup_outbox_queue: RollupOutboxQueue::new(&ROLLUP_OUTBOX_QUEUE_ROOT, max)
+                .unwrap(),
+        }
+    }
+
+    pub fn queue_len(&self) -> u32 {
+        self.meta.queue_len
+    }
+
+    pub fn incr_queue_len(&mut self) {
+        self.meta.queue_len += 1;
+    }
+
+    pub fn max(&self) -> u32 {
+        self.meta.max
+    }
+
+    /// Flushes the outbox queue in the order of the rollup outbox queue
+    /// then the outbox queue snapshot. The outbox has a maximum capacity
+    /// of 100 messages per level. Messages that cannot be flushed in the
+    /// current level are enqueued into the rollup outbox queue for the
+    /// next flush.
+    pub fn flush(
+        &mut self,
+        host: &mut impl Runtime,
+        snapshot: OutboxQueueSnapshot,
+    ) -> usize {
+        // 1. Flush the existing outbox queue
+        let mut flushed_count = self.rollup_outbox_queue.flush_queue(host);
+
+        // 2. Flush the outbox queue snapshot if there is space in the outbox
+        let mut tezos_outbox_messages =
+            snapshot.0.into_iter().map(|message| message.into());
+
+        for message in tezos_outbox_messages.by_ref() {
+            match Self::write_outbox_message(host, &message) {
+                Ok(()) => {
+                    flushed_count += 1;
```

Comment:
Why increment `flushed_count` here? `message` was never stored in `rollup_outbox_queue`?

Reply:
The point here is that instead of enqueing messages from `OutboxQueueSnapshot`s in the rollup's outbox queue only to flush it in the next line, we write as many messages to the kernel outbox ourselves. This saves us some unnecessary writes. So we take into account such writes by incrementing`flushed_out`.